PLFM-9199 - Metadata tasks API and database layer

nickgros opened the pull request:
- Add API models and controllers for Metadata Tasks
- Add database layer & autowired test for Metadata Tasks

Still need to
- Add the manager & its unit tests
- Add implementation to SynapseJavaClient
- Add an integration test for the entire feature

nickgros committed:
PLFM-9199 - Metadata tasks API and database layer

- Add API models and controllers for Metadata Tasks
- Add database layer & autowired test for Metadata Tasks

Still need to
- Add the manager & its unit tests
- Add implementation to SynapseJavaClient
- Add an integration test for the entire feature

nickgros committed:
PLFM-9199 - Changes from LLM review

john-hill commented on lib/jdomodels/src/main/resources/schema/MetadataTask-ddl.sql:
```diff
@@ -0,0 +1,22 @@
+    PRIMARY KEY (`ID`),
+    CONSTRAINT UNIQUE (`DATA_TYPE`, `PROJECT_ID`),
+    CONSTRAINT `METADATA_TASK_PROJECT_FK` FOREIGN KEY (`PROJECT_ID`) REFERENCES `NODE` (`ID`) ON DELETE CASCADE,
+    CONSTRAINT `METADATA_TASK_UPLOAD_FOLDER_FK` FOREIGN KEY (`UPLOAD_FOLDER_ID`) REFERENCES `NODE` (`ID`) ON DELETE SET NULL,
```
Remove the constraints, 
We are likely to have many types in the future, if we store each type as JSON column then we can add more types without changing the schema.

john-hill commented on lib/jdomodels/src/main/resources/schema/MetadataTask-ddl.sql:
```diff
@@ -0,0 +1,22 @@
+CREATE TABLE IF NOT EXISTS `METADATA_TASK`
+(
+    `ID`               BIGINT                                                        NOT NULL,
+    `DATA_TYPE`        varchar(256) CHARACTER SET utf8mb4 COLLATE utf8mb4_0900_ai_ci NOT NULL,
```
will need a test for all 256 chars to ensure there is no silent truncation.

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/metadata/task/DBOMetadataTask.java:
```diff
@@ -0,0 +1,288 @@
+                    .setRecordSetId(rs.getLong(SqlConstants.COL_METADATA_TASK_RECORD_SET_ID))
+                    .setEtag(rs.getString(SqlConstants.COL_METADATA_TASK_ETAG))
+                    .setCreatedBy(rs.getLong(SqlConstants.COL_METADATA_TASK_CREATED_BY))
+                    .setCreatedOn(rs.getTimestamp(SqlConstants.COL_METADATA_TASK_CREATED_ON))
```
any getLong() needs to do the rs.wasNull().

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/metadata/task/MetadataTaskDaoImpl.java:
```diff
@@ -0,0 +1,201 @@
+
+    private static final RowMapper<MetadataTask> METADATA_TASK_ROW_MAPPER = (rs, rowNum) -> {
+        MetadataTask task;
+        MetadataTaskType objectType = MetadataTaskType.valueOf(rs.getString(COL_METADATA_TASK_TASK_TYPE));
```
if we go the context route the parser will handle this for you.

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/metadata/task/MetadataTaskDaoImpl.java:
```diff
@@ -0,0 +1,201 @@
+    public void deleteMetadataTask(String taskId) {
+        basicDao.deleteObjectByPrimaryKey(DBOMetadataTask.class, new SinglePrimaryKeySqlParameterSource(Long.parseLong(taskId)));
+    }
+
```
need a method that fetches just the etag with a 'SELECT FOR UPDATE'

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/metadata/task/MetadataTaskDaoImpl.java:
```diff
@@ -0,0 +1,201 @@
+    @Override
+    @WriteTransaction
+    public MetadataTask updateMetadataTask(Long userId, MetadataTask toUpdate) {
+        MetadataTask currentTask = getMetadataTask(toUpdate.getTaskId()).orElseThrow(() -> new NotFoundException("No metadata task exists with id: " + toUpdate.getTaskId()));
```
just fetch the etag (with update)

john-hill commented on lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/metadata/task/MetadataTaskDaoAutowireTest.java:
```diff
@@ -0,0 +1,286 @@
+    }
+
+    @Test
+    public void testCRUDRecordBased() {
```
parameterized tests will eliminate the duplicates.

john-hill commented on lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/metadata/task/MetadataTaskDaoAutowireTest.java:
```diff
@@ -0,0 +1,286 @@
+    }
+
+    @Test
+    public void testNoDuplicateDataTypeWithinProject() {
```
use long name that consumes all of 256s chars (not silent key truncation)

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/metadata/task/MetadataTaskDaoImpl.java:
```diff
@@ -0,0 +1,201 @@
+                + SqlConstants.COL_METADATA_TASK_MODIFIED_ON + " = NOW() "
+                + "WHERE " + SqlConstants.COL_METADATA_TASK_ID + " = ?";
+
+        jdbcTemplate.update(
```
I like the unique message check in the create.

john-hill commented on lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/metadata/FileBasedMetadataTask.json:
```diff
@@ -0,0 +1,18 @@
+    {
+      "$ref": "org.sagebionetworks.repo.model.metadata.MetadataTask"
+    }
+  ],
```
this will likely change with the context version of the tasks.

john-hill commented on lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/metadata/MetadataTask.json:
```diff
@@ -0,0 +1,45 @@
+    "modifiedBy": {
+      "type": "string"
+    },
+    "concreteType": {
```
will be in the context

john-hill commented on lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/migration/MigrationType.json:
```diff
@@ -494,6 +494,10 @@
 			"description": "Table that tracks grid replica patches."
 		},
+		{
+		  "name": "METADATA_TASK",
```
will rename

john-hill commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/service/ServiceProviderImpl.java:
```diff
@@ -244,4 +246,9 @@ public DrsService getDrsService(){
 		return drsService;
 	}
 
+    @Override
```
I avoid using the service provider.

john-hill reviewed (changes requested):
Looks good! Minor changes.

nickgros committed:
PLFM-9199 - code review changes

- Rename MetadataTask to CurationTask across the codebase
- Move MetadataTask concrete-type specific properties into new `CurationTask.taskProperties` field, serialize to JSON in the DB.
- When updating a CurationTask, query for etag using "SELECT ... FOR UPDATE"
- Remove usage of CurationTaskService in ServiceProvider
- Remove FK references to NODE table except for parent project info
